Fixes for ROS2 tests

Tadinu opened the pull request:
`RRROS2GameMode` 
* Fix Sim init time measurement add `BeginTimestampInSec`
* Add `NativeSpawnableClasses`, being configured directly in GameMode ctor instead of GameState, also auto-registering them
`ATurtlebotBurger` fix root comp issue which affect `GetActorTransform()`

`URRObjectUtils::CreateMeshComponent()` fix `bStationary` in init mesh comp

External ROS2 tests
* robot_spawn/remove -> entity_spawn/remove
* utils: add `wait_for_entity_moving()`, used by `test_robot_cmd_vel_subscribed`. Move `CmdVelPublisher` to `cmd_vel_publisher.py`
* `wait_for_spawned_entity()` add `in_entity_ref_frame` param
* Add `BasePublisher, BaseSubscriber` classes, inherited by CmdVel, Path publisher, ModelState, TF subscriber

Commit:
test_robot_spawn.py wait 10.0 after spawning before GetEntityState
,using MapOrigin0 as robot_ref_frame instead of assuming world as origin
robot_spawn/remove -> entity_spawn/remove
wait_for_spawned_entity() add in_entity_ref_frame param
cmd_vel_publisher cleanup
URRObjectUtils::CreateMeshComponent() fix bStationary in init mesh comp

Commit:
RRROS2GameMode add BeginTimestampInSec + log timestep + rtf upon change
Add NativeSpawnableClasses, also auto-registering them
UE_TWO_PI
ATurtlebotBurger fix root comp

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp:
```diff
@@ -11,12 +11,25 @@
 {
     DefaultPawnClass = ARRGhostPlayerPawn::StaticClass();
+
+    // Static spawnable classes
```
should avoid adding thses static classes in this base class for test.

Tadinu replied:
So basically, this is the place configuring spawnable entity classes, which is not only for test, but more to accommodate `SpawnEntity` srv. Otherwise, they have to be manually configured in all maps' Level Blueprint.

yuokamoto replied:
move to .ini

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp:
```diff
@@ -55,6 +59,17 @@ void ATurtlebotBurger::SetupBody()
+    Super::PreInitializeComponents();
+    if (IsDynamicRuntimeRobot() && DefaultRoot->GetAttachParent())
+    {
+        // NOTE: This requires [DefaultRoot] to have been attached to some NEW Root, by which there is no need to promote its child
```
when "[DefaultRoot] to have been attached to some NEW Root" happen?

Tadinu replied:
That only happens if the RootComp has been switched to a new one earlier, and thus parenting DefaultRoot. But this part is actually unrequired here due to TurtlebotBurger is statically-built class, so removed.

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Robots/Turtlebot3/TurtlebotBurger.cpp:
```diff
@@ -27,9 +27,13 @@ void ATurtlebotBurger::SetupBody()
     CasterBack = CreateDefaultSubobject<UStaticMeshComponent>(TEXT("CasterBack"));
 
-    if (RootComponent)
+    // NOTE: This is critical, otherwise GetTransform() always returns one of [DefaultRoot] as initial one at spawning due to its being just a non-physics scene component
```
- Do you mean this is a issue at the spawning time only?
- If this is a issue to general robot, shouldn't be part of base robot?

Tadinu replied:
Yeah I added `SetBaseMeshComp()` for RRBaseRobot, so child classes, if creating a custom base mesh comp, could use to automatically make it new RootComp.

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Core/RRROS2GameMode.cpp:
```diff
@@ -113,6 +127,7 @@ void ARRROS2GameMode::SetFixedTimeStep(const float InStepSize)
     }
     FApp::SetUseFixedTimeStep(true);
     FApp::SetFixedDeltaTime(InStepSize);
+    UE_LOG_WITH_INFO_SHORT(LogRapyutaCore, Display, TEXT("FIXED TIMESTEP UPDATED: %f"), InStepSize);
```
Better to use Capital only beginning of word for consistency

Commit:
log case

Commit:
RRBaseRobot add BaseMeshComp + SetBaseMeshComp()
auto replacing DefaultRoot if needed

Commit:
Merge branch 'devel' into UpdateTestScript

Commit:
RRROS2GameMode add BPSpawnableClassNames, NativeSpawnableClasses to (Config)
, being configurable live in INI

Commit:
Missed !

Commit:
RRAssetUtils add FindClassFromPathName()
Used to automaticall fetch UClass* to act as TSubclassOf<> for spawnable entity models
Move all configs of BPSpawnableClassNames, NativeSpawnableClassPaths to INI

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp:
```diff
@@ -185,6 +185,32 @@ void ARRBaseRobot::PostInitializeComponents()
 
+void ARRBaseRobot::SetBaseMeshComp(UMeshComponent* InBaseMeshComp, bool bInMakeAsRoot, bool bInDestroyDefaultRoot)
+{
+    ensure(InBaseMeshComp);
```
should return if InBaseMeshComp == nullptr?

Commit:
only register valid configured entity type in INI

Commit:
reject invalid InBaseMeshComp

Commit:
Content/SkeletalRobots/Turtlebot3 add prefix to avoid naming conflict with static mesh-based TB3
base_subscriber.py no need for Rate, use time_step instead

yuokamoto commented on Content/SkeletalRobots/Turtlebot3/BallCasterSphereWheeled/BP_BallCasterSphereWheeledTurtlebotBurger.uasset:
is this just renamed? 
is this related to the test?

Tadinu replied:
yes, there is a bug related to searching for `BP_TurtlebotBurger` as defined in `BPSpawnableClassNames`, each of which must hold a unique name, so the spawning could find the correct class. They should always be regardless anw.

yuokamoto commented on ExternalTest/rr_sim_tests/rr_sim_tests/utils/path_publisher.py:
```diff
@@ -18,70 +18,39 @@
+                 in_robot_path:Path = Path()):
+        super().__init__(in_robot_namespace, in_topic_name, in_publish_num, in_publish_freq)
+        self._publisher = self.create_publisher(
+            Path, self._full_topic_name, qos_profile=PUBLISHER_DEFAULT_QOS
```
better to use in_topic_name directly. 
Name resolution should be done by ROS

yuokamoto commented on ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_subscriber.py:
```diff
@@ -0,0 +1,76 @@
+
+        # https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_impl.hpp
+        # https://github.com/ros2/rclcpp/issues/1767
+        self._full_topic_name = None
```
if this is a common feature for both pub and sub, should be separated

Tadinu replied:
You meant grouped?, I so added `base_pub_sub_node.py`

yuokamoto commented on ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py:
```diff
@@ -0,0 +1,143 @@
+        # https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_impl.hpp
+        # https://github.com/ros2/rclcpp/issues/1767
+        self._full_topic_name = None
+        if in_topic_name.startswith('/') or in_topic_name.startswith('~'):
```
~ means private name space, i.e. it become nodename/topic name

yuokamoto replied:
Does get_namespace include node name as well?
topic name should become ns/node_name/topic_name

Tadinu replied:
So it should be `get_effective_namespace()` in case of ~ I guess

Commit:
add base_pub_sub_node.py

yuokamoto reviewed (approved):
@chalice-graeme please review

Commit:
print() -> logging or node.get_logger()
